src/mgm.rs

```diff
@@ -319,111 +446,84 @@ impl MgmKey {
         Ok(())
     }
 
-    /// Encrypt with 3DES key
-    pub(crate) fn encrypt(&self, input: &[u8; DES_LEN_DES]) -> [u8; DES_LEN_DES] {
-        let mut output = input.to_owned();
-        TdesEde3::new(GenericArray::from_slice(&self.0))
-            .encrypt_block(GenericArray::from_mut_slice(&mut output));
-        output
+    /// Return the size of the key used in management operations for a given algorithm
+    pub const fn key_size(&self) -> u8 {
+        C::KEY_SIZE
     }
 
-    /// Decrypt with 3DES key
-    pub(crate) fn decrypt(&self, input: &[u8; DES_LEN_DES]) -> [u8; DES_LEN_DES] {
-        let mut output = input.to_owned();
-        TdesEde3::new(GenericArray::from_slice(&self.0))
-            .decrypt_block(GenericArray::from_mut_slice(&mut output));
-        output
+    /// Given a challenge from a card, decrypt it and return the value
+    pub(crate) fn card_challenge(&self, challenge: &[u8]) -> Result<Vec<u8>> {
+        if challenge.len() != C::block_size() {
+            return Err(Error::SizeError);
+        }
+
+        let mut output = challenge.to_owned();
+
+        C::new(&self.key).decrypt_block(GenericArray::from_mut_slice(&mut output));
+
+        Ok(output)
+    }
+
+    /// Checks the authentication matches the challenge and auth data
+    pub(crate) fn check_challenge(&self, challenge: &[u8], auth_data: &[u8]) -> Result<()> {
+        let mut response = challenge.to_owned();
+
+        if challenge.len() != C::block_size() {
+            return Err(Error::AuthenticationError);
+        }
+
+        C::new(&self.key).encrypt_block(GenericArray::from_mut_slice(&mut response));
+
+        use subtle::ConstantTimeEq;
+        if response.ct_eq(auth_data).unwrap_u8() != 1 {
+            return Err(Error::AuthenticationError);
+        }
+
+        Ok(())
+    }
+
+    /// Return the ID used to identify the key algorithm with APDU packets
+    pub(crate) fn algorithm_id(&self) -> u8 {
+        C::ALGORITHM_ID
     }
 }
 
-impl AsRef<[u8; DES_LEN_3DES]> for MgmKey {
-    fn as_ref(&self) -> &[u8; DES_LEN_3DES] {
-        &self.0
+impl<C: MgmKeyAlgorithm> AsRef<[u8]> for MgmKey<C> {
+    fn as_ref(&self) -> &[u8] {
+        self.key.as_ref()
     }
 }
 
 /// Default MGM key configured on all YubiKeys
-impl Default for MgmKey {
+impl<C: MgmKeyAlgorithm> Default for MgmKey<C> {
     fn default() -> Self {
-        MgmKey([
-            1, 2, 3, 4, 5, 6, 7, 8, 1, 2, 3, 4, 5, 6, 7, 8, 1, 2, 3, 4, 5, 6, 7, 8,
-        ])
+        let key = Key::<C>::from_slice(&DEFAULT_MGM_KEY).clone();
+        Self {
+            key,
+            _cipher: std::marker::PhantomData,
+        }
```

Comment:
Hi! I wanted to test your PR locally, but noticed that here you have Default implementation for all MgmKey. However, the `DEFAULT_MGM_KEY` is only 24 bytes, so the `Key::<C>::from_slice` will panic if someone tries to create `MgmKeyAes128::default()` or `MgmKeyAes256::default()`. Maybe it makes sense to implement Default only for 3DES and AES-192?

Reply:
Might be worth making default work regardless, although this might be an indication the API is not quite right yet as per the previous conversations.